sql/src/main/java/io/crate/planner/consumer/MultiSourceGroupByConsumer.java

```diff
@@ -0,0 +1,220 @@
+                orderBy,
+                limits.offset(),
+                limits.finalLimit(),
+                outputs
+            );
+            plan.addProjection(postAggregationProjection, null, null, outputs.size(), null);
+        }
+
+        /**
+         * Adds the group protection to the given plan in order to handle the groupBy.
+         */
+        private void addGroupProtection(Plan plan,
+                                        SplitPoints splitPoints,
+                                        List<Symbol> groupKeys) {
+            ProjectionBuilder projectionBuilder = new ProjectionBuilder(functions);
+
+            GroupProjection groupProjection = projectionBuilder.groupProjection(
+                splitPoints.toCollect(),
+                groupKeys,
+                splitPoints.aggregates(),
+                AggregateMode.ITER_FINAL,
+                RowGranularity.CLUSTER
+            );
+            plan.addProjection(groupProjection, null, null, groupProjection.outputs().size(), null);
+        }
+
+        /**
+         * Adds the filter protection to the given plan in order to handle the `having` clause.
+         */
+        private static void addFilterProtection(Plan plan,
+                                                List<Symbol> groupKeys,
+                                                SplitPoints splitPoints,
+                                                QuerySpec querySpec) {
+            Optional<HavingClause> havingClause = querySpec.having();
+
+            if (havingClause.isPresent()) {
+                List<Symbol> postGroupingOutputs = new ArrayList<>(
+                    groupKeys.size() +
+                    splitPoints.aggregates().size());
+                postGroupingOutputs.addAll(groupKeys);
+                postGroupingOutputs.addAll(splitPoints.aggregates());
+                HavingClause having = havingClause.get();
+                FilterProjection filterProjection = ProjectionBuilder.filterProjection(postGroupingOutputs, having);
+                plan.addProjection(filterProjection, null, null, null, null);
+            }
+        }
+
+        /**
+         * Remove limit, offset, order by and group by from RelationSource and MultiSourceSelect QuerySpec.
+         */
+        private static void removePostGroupingActionsFromQuerySpec(MultiSourceSelect mss, SplitPoints splitPoints) {
+            QuerySpec querySpec = mss.querySpec();
+            List<Symbol> outputs = Lists2.concatUnique(splitPoints.toCollect(), extractFieldsFromJoinConditions(mss));
+            querySpec.outputs(outputs);
+            querySpec.hasAggregates(false);
+            removePostGroupingActions(querySpec);
+
+            for (RelationSource relationSource : mss.sources().values()) {
+                removePostGroupingActions(relationSource.querySpec());
+            }
+        }
+
+        private static List<Field> extractFieldsFromJoinConditions(MultiSourceSelect statement) {
```

Comment:
This method could be shared - the one in `MultiSourceAggregationConsumer` is exactly the same.